library/core/src/intrinsics.rs

```diff
@@ -1130,7 +1130,10 @@ extern "rust-intrinsic" {
     /// may lead to unexpected and unstable compilation results. This makes `transmute` **incredibly
     /// unsafe**. `transmute` should be the absolute last resort.
     ///
-    /// Transmuting pointers to integers in a `const` context is [undefined behavior][ub].
+    /// Transmuting pointers *to* integers in a `const` context is [undefined behavior][ub],
+    /// unless the pointer was originally created by transmuting *from* an integer.
+    /// (That includes this function specifically and helpers like [`invalid`][crate::ptr::invalid],
+    /// but also semantically-equivalent conversions such as punning through `repr(C)` union fields.)
```

Comment:
```suggestion
    /// Transmuting pointers *to* integers in a `const` context is [undefined behavior][ub],
    /// unless the pointer was originally created *from* an integer.
    /// (That includes this function specifically, integer-to-pointer casts, and helpers like [`invalid`][crate::ptr::invalid],
    /// but also semantically-equivalent conversions such as punning through `repr(C)` union fields.)
```
We don't want to say that `invalid` is a transmute. We just want to say that `transmute(ptr::invalid(N)): *mut T` is allowed, as is `transmute(Nusize): *mut T`.

Reply:
Wait, does "that includes" refer to "created by" or to "transmuting to"? It must be "created by" as otherwise `invalid` doesn't work, but then this doesn't cover union field punning from `invalid` to an integer.

IMO this section shouldn't talk specifically about that anyway, that should fall out of the general principle that `transmute` and union field punning are equivalent.

Reply:
Maybe I should mention only [`invalid`][crate::ptr::invalid] explicitly?

Something like:
```
unless the pointer was originally converted *from* an integer.
(for example with this function or [`invalid`][crate::ptr::invalid])
```

Reply:
I'm pretty sure we want to allow transmuting `16 as *const u8` back to an integer. There is really no reason not to.

Reply:
The parenthetical explains what it means to convert from an integer. So then if you think casting is converting from an integer, it should be listed. :)